src/expr.rs

```diff
@@ -2273,9 +2342,34 @@ pub(crate) fn is_method_call(expr: &ast::Expr) -> bool {
     }
 }
 
+struct FloatSymbolParts<'a> {
+    integer_part: &'a str,
+    fractional_part: Option<&'a str>,
+    exponent: Option<&'a str>,
+}
+
+// Parses a float literal. The `symbol` must be a valid floating point literal without a type
+// suffix. Otherwise the function may panic or return wrong result.
+fn parse_float_symbol(symbol: &str) -> FloatSymbolParts<'_> {
+    // This regex may accept invalid float literals (such as `1`, `_` or `2.e3`). That's ok.
+    // We only use it to parse literals whose validity has already been established.
+    let float_literal_regex = static_regex!(r"^([0-9_]+)(?:\.([0-9_]+)?)?([eE][+-]?[0-9_]+)?$");
+    let caps = float_literal_regex.captures(symbol).unwrap();
+    FloatSymbolParts {
+        integer_part: caps.get(1).unwrap().as_str(),
+        fractional_part: caps.get(2).map(|m| m.as_str()),
+        exponent: caps.get(3).map(|m| m.as_str()),
+    }
+}
+
+fn is_zero_integer_literal(symbol: &str) -> bool {
+    let zero_literal_regex = static_regex!(r"^[0_]+$");
+    zero_literal_regex.is_match(symbol)
+}
```

Comment:
Would it be better to define this as a method on `FloatSymbolParts`? Can you also add a doc comment to help me better understand what this is used for.